modules/etcd/etcd.go

```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		return nil, fmt.Errorf("configure cluster: %w", err)
+	}
+
+	// configure CMD with the nodes
+	genericContainerReq.Cmd = configureCMD(settings)
+
+	if settings.clusterNetwork != nil {
+		// apply the network to the current node
+		err := tcnetwork.WithNetwork([]string{settings.nodeNames[settings.currentNode]}, settings.clusterNetwork)(&genericContainerReq)
+		if err != nil {
+			return nil, fmt.Errorf("with network: %w", err)
+		}
+	}
+
+	container, err := testcontainers.GenericContainer(ctx, genericContainerReq)
+	var c *EtcdContainer
+	if container != nil {
+		c = &EtcdContainer{Container: container}
+	}
+
+	if err != nil {
+		return c, fmt.Errorf("generic container: %w", err)
+	}
+
+	// only the first node creates the cluster
+	if settings.currentNode == 0 {
+		for i := 1; i < len(settings.nodeNames); i++ {
+			// move to the next node
+			clusterNode, err := Run(ctx, req.Image, append(clusterOpts, withCurrentNode(i))...)
+			if err != nil {
+				// return the parent cluster node and the error, so the caller can clean up.
+				return c, fmt.Errorf("run cluster node: %w", err)
+			}
+
+			c.childNodes = append(c.childNodes, clusterNode)
+		}
+	}
+
+	c.ClusterToken = settings.clusterToken
+
+	return c, nil
+}
+
+// configureCluster configures the cluster settings, ensuring that the cluster is properly configured with the necessary network and options,
+// avoiding duplicate application of options to be passed to the successive nodes.
+func configureCluster(ctx context.Context, settings *options, opts []testcontainers.ContainerCustomizer) ([]testcontainers.ContainerCustomizer, error) {
+	var clusterOpts []testcontainers.ContainerCustomizer
+	if len(settings.nodeNames) <= 1 {
+		return clusterOpts, nil
+	}
+
+	// pass cluster options to each node
+	etcdOpts := []Option{}
+	for _, opt := range opts {
+		// if the option is of type Option, it won't be applied to the settings
+		// this prevents the same option from being applied multiple times (e.g. updating the current node)
+		if apply, ok := opt.(Option); ok {
+			etcdOpts = append(etcdOpts, apply)
+		} else {
+			clusterOpts = append(clusterOpts, opt)
+		}
+	}
+
+	if settings.clusterNetwork == nil { // the first time the network is created
+		newNetwork, err := tcnetwork.New(ctx)
+		if err != nil {
+			return clusterOpts, fmt.Errorf("new network: %w", err)
+		}
+
+		// set the network for the first node
+		settings.clusterNetwork = newNetwork
+
+		clusterOpts = append(clusterOpts, withClusterNetwork(newNetwork)) // save the network for the next nodes
+	}
+
+	// we finally need to re-apply all the etcd-specific options
+	clusterOpts = append(clusterOpts, withClusterOptions(etcdOpts))
+
+	return clusterOpts, nil
+}
+
+// configureCMD configures the etcd command line arguments, based on the settings provided,
+// in order to create a cluster or a single-node instance.
+func configureCMD(settings options) []string {
+	cmds := []string{"etcd"}
+
+	if len(settings.nodeNames) < 1 {
```

Comment:
question: do we want to allow a cluster of 1 node, doesn't really make sense?

Reply:
I think it could be the case you want a single etcd instance

Reply:
Single instance yes but not single node cluster?

Reply:
I don't know the semantical difference between being a one-node cluster Vs being one instance (e.g. one elasticsearch instance is the same as one-node elasticsearch cluster).

In our case, the code just detects there is one node/instance, and it returns a single container. When the user requests a cluster (more than one nodes), then we return a parent container with children. So I do not see the difference here.

I could picture the Run function returning just one container and an eventual RunCluster one that separates concerns. But this is the purpose of the `WithNodes` option, right?

Reply:
Yer I don't see the need for a with cluster, but what I saw was all the additional command line options complexity for cluster setup which could be avoided for a single node. I believe the only reason etcd supports a single node cluster is so you can add nodes at a later date, which doesn't seem like a scenario you would need for testing.

Reply:
suggestion: With the change to WithNodes, this can be changed as we can no longer create a single node cluster.
```suggestion
	if len(settings.nodeNames) == 0 {
```

Reply:
>  I believe the only reason etcd supports a single node cluster is so you can add nodes at a later date
Indeed. When I started the feature it was to support two scenarios:
- one etcd node
- multiple etcd nodes in a cluster

No more

Reply:
You raise a good point, do we even need cluster support, should we just keep a patch about and add it later if there is demand, that avoids the hassle of deprecation?

Reply:
I think so, I might be wrong. I guess someone might want to test their app worked correctly if they killed a cluster node, but that would be an edge case and need additional work to expose the nodes themselves.

Reply:
@eddumelendez @kiview what are your thoughts about feature parity across modules in the libraries? Should we start simple with no cluster support, or just keep what this PR does?